src/main/java/dk/cngroup/trainings/spring/springassignment/service/AnimalServiceImpl.java

```diff
@@ -1,80 +1,109 @@
-                return this.addAnimal(animal);
-            }
-        }
-        else {
-            return null; //TODO dont return null
-        }
-    }
+	@Override
+	public Optional<Animal> addAnimal(Animal animal) {
+		if (AnimalValidationService.isValid(animal)) {
+			animal.setId(IdGenerator.getId());
+			return Optional.ofNullable(animalRepository.save(animal));
+		} else {
+			return Optional.empty();
+		}
+	}
 
-    @Override
-    public boolean deleteAnimalById(long id) {
-        if(animalRepository.existsById(id)) {
-            animalRepository.deleteById(id);
-            return true;
-        }else
+	@Override
+	public boolean deleteAnimalById(long id) {
+		if (animalRepository.existsById(id)) {
+			animalRepository.deleteById(id);
+			return true;
+		} else
 
-            return false;
-    }
+		{
+			return false;
+		}
+	}
 
-    @Override
-    public Optional<Animal> updateAnimalById(long id, @Valid Animal animal) { //TODO delete @Valid
-        if(isValid(animal) && animalRepository.existsById(id)) {
-            // Change the updated animal id and override it in the database.
-            animal.setId(id);
-            return Optional.ofNullable(animalRepository.save(animal));
-        } else {
-            return null; //TODO Optional.empty();
-        }
-    }
+	@Override
+	public Optional<Animal> updateAnimalById(long id, Animal animal) {
+		if (AnimalValidationService.isValid(animal) && animalRepository.existsById(id)) {
+			// Change the updated animal id and override it in the database.
+			animal.setId(id);
+			return Optional.ofNullable(animalRepository.save(animal));
+		} else {
+			return Optional.empty();
+		}
+	}
 
-    @Override
-    public boolean isValid(@Valid Animal animal) { //it would be possible to move this e.g. to AnimalValidationService
-        //TODO check also nullability
-        return animal.getName().length()>=Animal.NAME_MINIMUM_SIZE
-                && animal.getDescription().length()<Animal.DESCRIPTION_MAXIMUM_SIZE
-                && !animal.getDescription().toLowerCase().contains("penguin");
-    }
+	@Override
+	public Optional<CareTaker> addNewCareTakerToExistingAnimal(long id, CareTaker careTaker) {
+		Optional<Animal> animal = animalRepository.findById(id);
+		if (CareTakerValidationService.isValid(careTaker)) {
+			careTaker.setId(IdGenerator.getId());
+			List<Animal> animals = Arrays.asList(animal.get());
+			careTaker.setAnimals(animals);
+			CareTaker addedCareTaker = careTakerRepository.save(careTaker);
+			return Optional.ofNullable(addedCareTaker);
+		} else {
+			return Optional.empty();
+		}
+	}
+
+	@Override
+	public String addExistingCareTakerToExistingAnimal(long animalId, long careTakerId) {
+		Optional<Animal> animal = animalRepository.findById(animalId);
+		Optional<CareTaker> careTaker = careTakerRepository.findById(careTakerId);
+		if (!animal.isPresent()) {
+			return "Failed: Animal id doesn't exist";
```

Comment:
It would be definately better to Throw exceptions here then return some String. Imagine you are a client who wants to use this service, then you would need to parse the String from the return value to determine, what actually happened. And parsing is hard and also can contain mistakes and compiler won't help you. If you just throw exceptions, you need to handle them correctly, otherwise your app won't even start.

Reply:
Yeah, I didn't like returning plain hard coded strings. I will think of the exceptions option.